libnet_raw: don't change the TX buffer size for raw sockets

beni-sandu opened the pull request:
While doing some investigation around syscall usage on one of my apps that use libnet, I noticed the following piece of code that tries to configure the TX buffer size for raw sockets as close to 1MB as possible (the default for this on most Linux systems seems to be around 200K).

Because of this loop, for every libnet raw context initialization, there are over 6500 setsockopt syscalls generated, for no obvious reason.

Simple test on udp1.c sample, before change:
```
$ sudo strace -c -f ./udp1 -s 192.168.4.1 -d 192.168.4.2
Running with: libnet version 1.3-dev
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
Packets sent:  10
Packet errors: 0
Bytes written: 480
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 98.25    0.014458           2      6529           setsockopt
  1.21    0.000178           8        22           write
  0.48    0.000071           7        10           sendto
  0.06    0.000009           2         4           close
  0.00    0.000000           0         2           read
  0.00    0.000000           0         4           fstat
  0.00    0.000000           0        13           mmap
  0.00    0.000000           0         4           mprotect
  0.00    0.000000           0         1           munmap
  0.00    0.000000           0         3           brk
  0.00    0.000000           0         6           pread64
  0.00    0.000000           0         1         1 access
  0.00    0.000000           0         1           socket
  0.00    0.000000           0         1           getsockopt
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         2         1 arch_prctl
  0.00    0.000000           0         3           openat
------ ----------- ----------- --------- --------- ----------------
100.00    0.014716                  6607         2 total
```

After change:
```
$ sudo strace -c -f ./udp1 -s 192.168.4.1 -d 192.168.4.2
Running with: libnet version 1.3-dev
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
48 byte packet, ready to go
Wrote 48 byte UDP packet; check the wire.
Packets sent:  10
Packet errors: 0
Bytes written: 480
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 40.12    0.000067           6        10           sendto
 34.13    0.000057           2        22           write
 25.75    0.000043          10         4           close
  0.00    0.000000           0         2           read
  0.00    0.000000           0         4           fstat
  0.00    0.000000           0        13           mmap
  0.00    0.000000           0         4           mprotect
  0.00    0.000000           0         1           munmap
  0.00    0.000000           0         3           brk
  0.00    0.000000           0         6           pread64
  0.00    0.000000           0         1         1 access
  0.00    0.000000           0         1           socket
  0.00    0.000000           0         2           setsockopt
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         2         1 arch_prctl
  0.00    0.000000           0         3           openat
------ ----------- ----------- --------- --------- ----------------
100.00    0.000167                    79         2 total
```

Can anyone let me know about the history behind this? Can we drop this piece of code as it seem to generate a lot of unneeded syscalls?

Thanks,
Beni

troglobit commented:
Good question, as co-maintainer I'm also curious.  At first glance this is something that should probably be handled by the application using libnet rather than implicit in the library.

Maybe @sam-github can shed some light on this?

sam-github commented:
Sorry, predates me. Seems less than ideal, like something that should be setup once, and left alone. setting buffer size might be system dependent, maybe an api wrapper is needed for it, but setting it on every send seems excessive to me, too.

troglobit commented:
Should we drop it then, as the PR suggests?

sam-github commented:
as far as this PR goes, its been a while, but it seems the relevant code that is being deleted might be worth factoring out into a standalone API.

troglobit commented:
That's reasonable, care to submit a follow-up on that note @beni-sandu?

beni-sandu commented:
Sure, I just don't know about the arbitrary number in there.

Is it good enough to wrap this chunk into something like **libnet_setsock_max_txbufsize()** that does exactly the same thing?
Is there some naming convention that needs to be followed? Or maybe a more specific API that can take the desired TX buffer size as input?

beni-sandu commented:
I updated the commit, moved this to an internal wrapper called **libnet_setsock_max_txbufsize()** that can be used and does exactly the same thing. I kept the original ifdef and the 1MB value for the moment.

Just a personal observation, having a quick look through the code, there are a lot of parts (especially remnants of the past) that seem they could use some serious cleanup...

Anyway, please let me know if this is good enough or it needs any adjustments.

troglobit commented:
First of all, you've changed the patch to refactor the `SO_BROADCAST` and `SO_BINDTODEVICE` instead of the `SO_SNDBUF` sockopt, what gives?

Second, yes there is code that needs cleanup. You can open a discussion about specific parts so we can agree on both general and specific approaches to refactoring.

beni-sandu commented:
Hey @troglobit , my apologies if this came out the wrong way, I was not pointing fingers at anyone, just saying that this piece of code is an example of old stuff that is around without much information on context (at least for me, as I am not familiar with libnet code in general, just used it for a bit).

I have moved the SO_SNDBUF chunk below the initial function (libnet_finish_setup_socket), so because of the position in the file, the diff makes it look like I changed SO_BROADCAST.

https://github.com/beni-sandu/libnet/blob/master/src/libnet_raw.c#L68

troglobit commented:
No worries, I just wanted to be clear, I sometimes come off as too harsh which is not my intention at all.  Actually, I'm in favor of dropping this piece of code altogether, but @sam-github seems to want to keep it, and he's the boss :-)

The diff looked really weird, thank you for clearing this up for me!  I'm thinking we should probably move this helper function to libned_adv.c or libnet_init.c and rename it.  How about just below `libnet_getfd()` in `libnet_init.c`, and renaming it to `libnet_setfd_max_sndbuf()` instead?

Also, if we're keeping this maybe add a `, int max_bytes` parameter to the function definition?  

The `SO_SNDBUF` sockopt in Linux is weird:
> Sets  or  gets  the  maximum  socket send buffer in bytes.  The kernel doubles this value (to allow space for
              bookkeeping overhead) when it is set using setsockopt(2), and this doubled  value  is  returned  by  getsock‐
              opt(2).   The  default value is set by the /proc/sys/net/core/wmem_default file and the maximum allowed value
              is set by the /proc/sys/net/core/wmem_max file.  The minimum (doubled) value for this option is 2048.

beni-sandu commented:
Yeah, I checked on a few Linux systems, and seems the advertised default value is also the max value in that procfs file, which is 212992.

If we take max_bytes as argument, we should probably just try and set it once with setsockopt to that value, and return with -1 if it fails, instead of looping like that.

I am with you on this functionality being a bit out of scope for libnet, but I guess it can be dropped later if it really proves of no use.

I pushed a new patch, please have a look.

beni-sandu commented:
> I haven't looked at this, its possible that there is one time socket initialization that should have the buf expanded right after the socket is created? instead of before every send? That would be backwards compatible, which is always comforting.

That is actually what's happening right now, at the end of every raw socket init, the loop is trying to configure the buffer as close to 1MB as possible, it doesn't happen on every send. The thing that makes me sad is that on a Linux system with the default 200K buffer, this generates ~6500 extra setsockopt syscalls for every raw socket init. I was hoping we can drop it and let the user handle setting a desired bufsize if needed, but of course it's up to you guys.

troglobit reviewed (changes requested):
Awesome work on this so far! Could you please also update the ChangeLog.md file with a human-readable summary of the change?  Then I'm happy and we can merge :)

beni-sandu commented:
Seems github closed the PR automatically while I was updating my top commit.

I have added a comment in the Changelog, @troglobit please have a look when you get the chance: https://github.com/beni-sandu/libnet/commit/79ff87bd02d21379c33496c1fa50a89312773140 and let me know if there is anything else that is needed.

Maybe you can just pull/cherry-pick that one if everything is okay so I don't have to reopen the PR?

troglobit commented:
Thank you!  I'll probably reword the changelog entry later on, but it's a great placeholder that summarizes the change.  

Have no idea why GitHub closed the PR when you pushed, but the changelog commit is now part of the PR and I'll merge as soon as all the builders/checks have completed.

beni-sandu commented:
Great, thanks for the support @troglobit and @sam-github .
